File: week-4/valid-parenthesis/valid-parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pairs = { '}': '{', ']': '[', ')': '(' }
        for i in s:
            if i in pairs:
                if stack and stack[-1] == pairs[i]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(i)
            
        return len(stack) == 0

File: week-4/valid-parenthesis/valid-parenthesis.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    const pairs = { ']': '[', ')': '(', '}': '{' };
    const stack = [];
    
    for (let i of s) {
        if (i in pairs) {
            if (stack.length != 0 && stack[stack.length-1] === pairs[i])
                stack.pop();
            else
                return false;
        } else
            stack.push(i);
    }
    
    return stack.length === 0;
};
